### PR TITLE
MiqSshUtil updates

### PR DESCRIPTION
**Pull request by djberg96**
This PR fixes a bug in the `MiqSshUtil#exec` method, and makes a few tweaks to the constructor and exec method in general.

Fixes:

Error handling is currently semi-busted. The `on_extended_data` hook is missing a yield param. The net result is that there is never any data in the error output. In addition, the `status` check could break since it defaults to nil, with the result that it will try to call `.zero?` on a nil object, which will raise an exception. So, I changed the default from nil to 0.

Updates:

* Some unused code that was commented out was removed, i.e. logging.
* Some workaround code for older versions of net-ssh that is no longer relevant was removed.
* You can now configure the `use_agent` option. It still defaults to false for backwards compatibility.
* The error handling was updated slightly. It now no longer relies on activesupport methods, and checks the error buffer, too.

You can see these issues in practice with this simple script:

```
# Assuming you're in the manageiq-gems-pending repo
$:.unshift('lib/gems/pending/util')
require 'MiqSshUtil'
require 'etc'

# Adjust as desired
hostname = 'localhost'
username = Etc.getlogin
password = 'xxxxxxx'
command  = 'bogus'

pubkey = IO.read(File.join(Dir.home, '.ssh', 'id_rsa.pub'))

# You will have to edit MiqSshUtil.rb if you're on master and need 'use_agent'.
opts = {
  :key_data => pubkey,
  :use_agent => true
}

ssh = MiqSshUtil.new(hostname, username, password)

begin
  result = ssh.exec(command)
rescue => err
  puts "OOPS: #{err.message}"
else
  puts "RESULT: #{result}"
end
```

BEFORE:

`RESULT: ` <---- Blank! And not registered as an error!

AFTER:

`OOPS: MiqSshUtil::exec - Command 'bogus' failed: bash: bogus: command not found, status: 127`

Partially addresses https://bugzilla.redhat.com/show_bug.cgi?id=1719689

### CODE REVIEW

**Comment by djberg96** [comment]
@roliveri How's it look?

**Comment by djberg96** [comment]
@miq-bot add_reviewer @bdunne

**Comment by djberg96** [comment]
@miq-bot add_label bug, changelog/yes

**Comment by djberg96** [comment]
@miq-bot add_label hammer/yes

File: lib/gems/pending/util/MiqSshUtil.rb [hunk 1]
```diff
@@ -149,18 +139,18 @@ def exec(cmd, doneStr = nil, stdin = nil)
 
         channel.on_close do |_channel|
           $log.debug "MiqSshUtil::exec - Command: #{cmd}, exit status: #{status}" if $log
-          unless signal.nil? || status.zero?
-            raise "MiqSshUtil::exec - Command #{cmd}, exited with signal #{signal}" unless signal.nil?
-            raise "MiqSshUtil::exec - Command #{cmd}, exited with status #{status}" if errBuf.empty?
-            raise "MiqSshUtil::exec - Command #{cmd} failed: #{errBuf}, status: #{status}"
+          if !signal.nil? || status.nonzero? || !errBuf.empty?
```
Comment (carbonin) [inline lib/gems/pending/util/MiqSshUtil.rb]: I think the logic here is being changed.

Previously we would enter the block only `if signal != nil && status != 0`

Now we will enter the block if only `signal != nil`

I think the conditional you want here is:
```ruby
if !errBuf.empty? || (!signal.nil? && status.nonzero?)
```

Alternatively you could change a bit less and go with:
```ruby
if !(signal.nil? || status.zero?) || !errBuf.empty?
```

As a side note it seems like the content of the block is checking against the possibility that `signal` could still be nil even though it has to be non-nil in the original implementation. So if the conditional was wrong to begin with then this looks good. 

I'm just not sure which direction you were trying to go with this.

-----------------------------------------

FWIW this is why I don't like `unless` with more than one condition.

Comment (djberg96) [reply lib/gems/pending/util/MiqSshUtil.rb]: @carbonin Not sure what you mean:

`if !signal.nil? || status.nonzero? || !errBuf.empty?`

We'll enter the block if any of those conditions is true.

Comment (Fryguy) [reply lib/gems/pending/util/MiqSshUtil.rb]: @djberg96 You applied the unless wrong, I think...

Previously...

```ruby
unless signal.nil? || status.zero?
# means
if !(signal.nil? || status.zero?)
# =>
if !signal.nil? && !status.zero?
# =>
if !signal.nil? && status.nonzero?
```

However you have (ignoring the new extra thing)

```ruby
if !signal.nil? || status.nonzero?
```

File: lib/gems/pending/util/MiqSshUtil.rb [hunk 2]
```diff
@@ -149,18 +139,18 @@ def exec(cmd, doneStr = nil, stdin = nil)
...
+            raise "MiqSshUtil::exec - Command '#{cmd}' failed: #{errBuf.chomp}, status: #{status}"
           end
           return outBuf
         end
 
         $log.debug "MiqSshUtil::exec - Command: #{cmd} started." if $log
         channel.exec(cmd) do |chan, success|
           raise "MiqSshUtil::exec - Could not execute command #{cmd}" unless success
-          if stdin.present?
+          unless stdin.nil?
```
Comment (Fryguy) [inline lib/gems/pending/util/MiqSshUtil.rb]: Why change this?  The present? was clearer than the new double negative, IMO.

Comment (Fryguy) [reply lib/gems/pending/util/MiqSshUtil.rb]: If you're dropping activesupport, then perhaps, just `if stdin`?

File: lib/gems/pending/util/MiqSshUtil.rb (hunk 1 above)
Comment (djberg96) [reply lib/gems/pending/util/MiqSshUtil.rb]: @Fryguy right, I'm saying the logic I have now is the logic I want. If any of those conditions is met then something went wrong.

File: lib/gems/pending/util/MiqSshUtil.rb (hunk 2 above)
Comment (djberg96) [reply lib/gems/pending/util/MiqSshUtil.rb]: Alright, I'll have to add an explicit require line for require 'active_support/core_ext/object/blank', though.

Comment (djberg96) [reply lib/gems/pending/util/MiqSshUtil.rb]: @Fryguy Updated. :)

File: lib/gems/pending/util/MiqSshUtil.rb (hunk 1 above)
Comment (carbonin) [reply lib/gems/pending/util/MiqSshUtil.rb]: @djberg96 I was just pointing out that the logic changed. You didn't call out this change in the description or commit message so I didn't know if it was intentional.

I know specs for this are probably tough to write, but having those would clarify intent quite nicely.

**Comment by djberg96** [comment]
@carbonin Alright, I've added some specs that should demonstrate that it behaves as expected. I didn't want to go down this road originally because I knew it would mean making a couple other refactorings.

Namely, the `Channel#exec` block had to be moved up (so that channel failure could be checked before command failure), and I had to strip the signal check result because it can (apparently) return a blank string of more than 1 character.

**Comment by simaishi** [comment]
Hammer backport details:

```
$ git log -1
commit 7300bf53e17784a7c12dc1133239bc8b76a61f8a
Author: Jason Frey <jfrey@redhat.com>
Date:   Tue Jul 2 16:16:08 2019 -0400

    Merge pull request #437 from djberg96/miq_ssh_util_updates
    
    MiqSshUtil updates
    
    (cherry picked from commit 02c9fb67797331640e2840682d8234f2f4a5415c)
    
    Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1726439
```

**Pull request merged**